tsdb/shard.go

```diff
@@ -1302,6 +1306,145 @@ func (a Shards) FieldKeysByPredicate(expr influxql.Expr) (map[string][]string, e
 	return all, nil
 }
 
+// consecutiveAndChildren finds all child nodes of consecutive
+// influxql.BinaryExpr with AND operator nodes ("AND nodes") which are not
+// themselves AND nodes. This may be the root of the tree if the root of the
+// tree is not an AND node.
+type consecutiveAndChildren struct {
+	children []influxql.Node
+}
+
+func (v *consecutiveAndChildren) Visit(node influxql.Node) influxql.Visitor {
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		if n.Op == influxql.AND {
+			return v
+		}
+	case *influxql.ParenExpr:
+		// Parens are essentially a no-op and can be traversed through.
+		return v
+	}
+
+	// If this wasn't a BinaryExpr with an AND operator or a Paren, record this
+	// child node and stop the search for this branch.
+	v.children = append(v.children, node)
+	return nil
+}
+
+// orMeasurementTree determines if a tree (or subtree) represents a grouping of
+// exclusively measurement names OR'd together with EQ operators for the
+// measurements themselves. It collects the list of measurement names
+// encountered and records the validity of the tree.
+type orMeasurementTree struct {
+	measurementKey   string
+	measurementNames []string
+	valid            bool
+}
+
+func (v *orMeasurementTree) Visit(node influxql.Node) influxql.Visitor {
+	// A measurement occuring in a tree which has been proven to be invalid
+	// results in the measurement optimization not applying, so the evaluation can
+	// be stopped early in this case.
+	if !v.valid && len(v.measurementNames) > 0 {
+		return nil
+	}
+
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		if n.Op == influxql.OR {
```

Comment:
I don't think this is properly handling things like `r._name = "m1" || r.bool_col_value`

Reply:
I'm not sure if a query like that could be expressed as an InfluxQL predicate and thus probably wouldn't end up here - but I went ahead and added a check to verify that the LHS and RHS are both either a binary expression or a paren that should address this.